test/e2e_node/device_plugin_failures_test.go

```diff
@@ -0,0 +1,247 @@
+
+		// at first all the devices are healthy
+		gomega.Eventually(getNodeAllocatableValue, f.Timeouts.NodeStatusUpdate, f.Timeouts.Poll).WithArguments(resourceName).Should(gomega.Equal(4))
+
+		// now make one device unhealthy
+		devices[3].Health = kubeletdevicepluginv1beta1.Unhealthy
+		plugin.UpdateDevices(devices)
+
+		gomega.Eventually(getNodeAllocatableValue, f.Timeouts.NodeStatusUpdate, f.Timeouts.Poll).WithArguments(resourceName).Should(gomega.Equal(3))
+
+		// now make the device healthy again
+		devices[3].Health = kubeletdevicepluginv1beta1.Healthy
+		plugin.UpdateDevices(devices)
+
+		gomega.Eventually(getNodeAllocatableValue, f.Timeouts.NodeStatusUpdate, f.Timeouts.Poll).WithArguments(resourceName).Should(gomega.Equal(4))
+
+		// now make two devices unhealthy
+		devices[1].Health = kubeletdevicepluginv1beta1.Unhealthy
+		devices[3].Health = kubeletdevicepluginv1beta1.Unhealthy
+		plugin.UpdateDevices(devices)
+
+		gomega.Eventually(getNodeAllocatableValue, f.Timeouts.NodeStatusUpdate, f.Timeouts.Poll).WithArguments(resourceName).Should(gomega.Equal(2))
+
+		// now make the device healthy again
+		devices[3].Health = kubeletdevicepluginv1beta1.Healthy
+		plugin.UpdateDevices(devices)
+
+		gomega.Eventually(getNodeAllocatableValue, f.Timeouts.NodeStatusUpdate, f.Timeouts.Poll).WithArguments(resourceName).Should(gomega.Equal(3))
+
+		// now make the device healthy again
+		devices[1].Health = kubeletdevicepluginv1beta1.Healthy
+		plugin.UpdateDevices(devices)
+
+		gomega.Eventually(getNodeAllocatableValue, f.Timeouts.NodeStatusUpdate, f.Timeouts.Poll).WithArguments(resourceName).Should(gomega.Equal(4))
+	})
+
+	ginkgo.It("when ListAndWatch fails immediately, node allocatable will be set to zero and kubelet will not retry to list resources", func(ctx context.Context) {
+		// randomizing so tests can run in parallel
+		resourceName := fmt.Sprintf("test.device/%s", f.UniqueName)
+		devices := []kubeletdevicepluginv1beta1.Device{{ID: "testdevice", Health: kubeletdevicepluginv1beta1.Healthy}}
+
+		// Initially, there are no allocatable of this resource
+		gomega.Eventually(getNodeAllocatableValue, f.Timeouts.NodeStatusUpdate, f.Timeouts.Poll).WithArguments(resourceName).Should(gomega.Equal(-1))
+
+		plugin := testdeviceplugin.NewDevicePlugin(func(name string) error {
+			if name == "ListAndWatch" {
+				return fmt.Errorf("ListAndWatch failed")
+			}
+			return nil
+		})
+
+		err := plugin.RegisterDevicePlugin(f.UniqueName, resourceName, devices)
+		defer plugin.Stop() // should stop even if registration failed
+		gomega.Expect(err).To(gomega.Succeed())
+
+		// kubelet registers the resource, but will not have any allocatable
+		gomega.Eventually(getNodeAllocatableValue, f.Timeouts.NodeStatusUpdate, f.Timeouts.Poll).WithArguments(resourceName).Should(gomega.Equal(0))
+
+		// kubelet will never retry ListAndWatch (this will sleep for a long time)
+		gomega.Consistently(plugin.Calls, 1*time.Minute, f.Timeouts.Poll).Should(gomega.HaveLen(2))
```

Comment:
arguably 1 minute is probably short time (I'd go for 2 or 3 minutes even, or at least consider) because we have no other signal than keep checking this value "long enough" and we longer we check the more certain we are.

Reply:
yes, but it makes iterating over tests soooo long. I will make it into a constant and fix up before removing the WIP

Reply:
great, works for me